Remove ObjectToImage

danieldresser-ie opened the pull request:
ObjectToImage is used only in the test cases, and we're working on gradually deprecating IECore::ImagePrimitive in favour of either OpenImageIO, or Gaffer's image processing.  The code to load from an ImagePrimitive hasn't been updated to deal with the deep flag, and John suggested instead of updating it, it could just be removed.

The only tricky bit is some places in the tests where we use ObjectToImage to load an image with a geometric pattern.  We now have a nicer way to generate geometric patterns: OSLImage.

However, now that I think about it, I'm not sure if I'm allowed to use OSLImage in the GafferImage test cases, even though it is by far the cleanest option.  I converted ImageSamplerTest to see what the change would look like, but I'll hold off on converting the other two test cases until you confirm if this is acceptable.

( I guess I should warn that I did also use OSL in order to set up one of the last DeepState tests I needed in order to confirm that that was all working ... )

( Also note that this is working on top of the Deep PR, so the history will be a total mess until that gets merged ).

johnhaddon commented:
Hey Daniel, thanks for taking this on...

> I'm not sure if I'm allowed to use OSLImage in the GafferImage test cases, even though it is by far the cleanest option.

Sorry, no. It may be clean in terms of having simple code in the test cases, but it's the opposite in terms of introducing unnecessary dependencies between modules. It's arguably _very_ thin in this case, but it's still the end of a wedge. This isn't just theoretical; at least one person in the real world has remarked that Gaffer's modularity allowed them to adopt modules piecemeal in an environment that otherwise had the potential for versionitis.

> I'll hold off on converting the other two test cases until you confirm if this is acceptable.

Thanks for holding off. In the case you've already done, I think a couple of Ramp nodes can be used to the same effect. For the others, I'm not averse to storing a small input image in the repo if it's the only remaining option. When you're doing this, could you rebase without all the deep commits, so we can merge this independently? Or is that a pain and we'd be better to merge after the deep?

danieldresser commented:
> This isn't just theoretical; at least one person in the real world has remarked that Gaffer's modularity allowed them to adopt modules piecemeal in an environment that otherwise had the potential for versionitis.

Hmm, still not sure I truly buy this.  It's great that this person can adopt Gaffer piecemeal without a core part, but do they truly need to run 100% test coverage?  I can certainly put in a skipIf so our hypothetical non-OSL user doesn't see a confusing failure if they do happen to run the tests, but is it really worth avoiding OSL altogether?

The sum-of-ramps trick probably should work in this case, but it's definitely less clear.  The tests actually are clearer when you can see the code for exactly what's in the reference image, rather than having to load up an image from disk and eyeball / colorpick it to figure out what you're testing against.

For the deep PR, I don't think there's currently any other way of creating a channel where all samples have their depth index, and I don't think we want to store a bunch of massive deep images for this purpose, so I guess I'd just throw out that piece of test and hope no one breaks it now that I've gotten it working.

As for commit history - it's going to be a conflict, so I need to include things one way or the other.  I was assuming that the deep PR was closer to merging than this

danieldresser-ie commented:
OK, while I couldn't find any good way to get rid of OSLImage in the deep stuff, this isn't too bad.  I use the double ramp trick, and some images written out to disk.

The remaining uses of ObjectToImage are pulling images out of the display server during interactive renders.  Do you think I can rewrite that to use a catalog?

johnhaddon commented:
> It's great that this person can adopt Gaffer piecemeal without a core part, but do they truly need to run 100% test coverage?... The tests actually are clearer when you can see the code for exactly what's in the reference image...

Yeah, those are reasonable points, and why I said it was arguably a _very_ thin end of a wedge. My concern is that experience shows that if you don't police boundaries actively, you can easily end up with a lot of unintentional cross-dependencies which make subsequent redeployment or refactoring hard.

> The remaining uses of ObjectToImage are pulling images out of the display server during interactive renders. Do you think I can rewrite that to use a catalog?

These are the various *RenderTest ones? Yep, that sounds reasonable.

danieldresser-ie commented:
I got one test in InteractiveArnoldRenderTest working using Catalogue, but it was more painful than expected.  I'm wondering if you have any ideas how to make this cleaner before I hit the many tests in the generic InteractiveRenderTest.

What I've pieced together:
* If there is no UIThreadCallHandler, the "bit sketchy" line here instantly crashes: https://github.com/GafferHQ/gaffer/blob/1fc556be0ecf99aef541a1510d633f5acb9b09b7/src/GafferImage/Display.cpp#L117
* I can't use a total no-op UI thread call handler, because the catalog actually relies on it to call driverCreated
* CatalogueTest uses with GafferTest.ParallelAlgoTest.ExpectedUIThreadCall(), but that expects exactly one UI thread call, and actually rendering with Arnold seems to generate a couple of these. 
* I don't want to count them exactly - I just want to render, then trigger any UI thread calls that have been generated, then check the catalogue for what's arrived.  The code in InteractiveArnoldRenderTest does this, but it's not exactly pretty.

johnhaddon commented:
> I'm wondering if you have any ideas how to make this cleaner before I hit the many tests in the generic InteractiveRenderTest.

Hmm, this UI thread emulation stuff has been a constant source of problems in the tests. Maybe we can wrap up what you've done into a little helper not dissimilar to `ParallelAlgoTest.ExpectedUIThreadCall`?

```
with ParallelAlgoTest.UIThreadCallHandler() as handler :

    script["render"]["state"].setValue( script["render"].State.Running )

    handler.waitFor( 1 ) # Replaces the sleep

    script["render"]["state"].setValue( script["render"].State.Stopped )
```

I think that might work, and also give us a single place to harden against some of the problems that could occur with the existing code :

- Protect against concurrent access to the list of things to do
- Handle calls continuously in `waitFor()` rather than calling `sleep()` and then doing everything at once. I think there _might_ be call patterns where the background thread can't post the next event until the first one is handled, meaning that the sleep approach might miss events?
- Drain the list of events in handler `__exit__`, since in our example, events might have piled up again before we stop the renderer.

How does that sound?

danieldresser-ie commented:
OK, this seems to be working.  Definitely take a look at my implementation of UIThreadCallHandler - I didn't really understand the complexity of ExpectedUIThreadCall, UIThreadCallHandler just uses a queue that's supposed to be threadsafe, which should take care of most concurrency issues, but there's probably some reason why ExpectedUIThreadCall is doing some more complicated looking stuff.

johnhaddon commented:
LGTM, but I'm obviously missing something because the CI is hanging on `GafferImageTest.CatalogueTest.CatalogueTest.testCacheReuse ()`. I presume this is something to do with the changes to UI call handling, but I'm not sure how.

> I didn't really understand the complexity of ExpectedUIThreadCall

From what I recall, some of the complexity comes from a need to nest `with ExpectedUIThreadCall` blocks. But I'm wondering now if we couldn't just add a `waitForOne()` method to your new class and use that everywhere. No need to address that right now though...

danieldresser-ie commented:
The hang in `testCacheReuse` does not appear relevant - I ran again, and that test ran fine ( although there were a few other failures, which I've now fixed ).

I've seen that test hanging fairly frequently locally when I have Gaffer open and have been using catalogs.  It's something to do with having catalogs open in other processes, but I don't think it's related to this PR - I've been seeing it for a while.

Interestingly, I was actually able to fairly consistently reproduce a hang in that test by running the test suite with an ImageEngine override that caused the Appleseed interactive tests to start rendering, and then fail before calling `rendere["state"].setValue( Stopped )`.  This doesn't appear to be the cause of the problem on Azure - there were no failures, so maybe we had 2 test processes at once on the same machine ... but maybe it would offer a more consistent way to figure out how the hang occurs.

Anyway, lets see if the tests pass now that I've fixed the other tidbits. If they do, it probably is safe to merge this, even if it is a good time for me to look into the likely unrelated testCacheReuse hang.

danieldresser commented:
Oops, must not have synced to the repo - the tests that are currently failing are the ones I fixed last thing Friday.

danieldresser-ie commented:
It appears that I've been cursed.  This PR is now showing a different, consistent hang, but only on Mac.  I haven't investigated this at all yet - it's a bit tricky when it's only on Mac.

What I have investigated is the hang in CatalogueTest.testCacheReuse that showed up once.  I don't have any idea why it failed on the test server, or why it fails occasionally when I have Gaffer open interactively while I'm running tests.  What I do now understand is why it hangs if I run `gaffer test` in one of my test envs.

In this test env, there's some IE stuff loaded that messes up the shader assignment, resulting in an exception being thrown, and halting a test.  What I've now realized is that this results in a failure that happens before calling the UI callback, meaning that Display::dataReceived has been called, but Display::dataReceivedUI has not been called.  This leaves the display in a bad state where dataReceivedUI will never be called, and any new calls to dataReceived will just add to the "current batch", which will just wait forever, because the job of executing the current batch falls to the code that originally ran and failed.

This feels fairly fragile - I still don't know where the sporadic failures are happening, but I do now understand how a sporadic failure could turn into a hang in a completely different part of the test suite.  Any thoughts on how to improve this?

Or thoughts on what's going on now with testEditContext failing on Mac?  I guess it could be related to my increased use of Catalogues and some sort of fragility of UI callbacks?

danieldresser-ie commented:
Replaced by #3511
